api/src/auth/api_gateway_key_auth.py

```diff
@@ -0,0 +1,78 @@
+import logging
+
+from sqlalchemy import select
+from sqlalchemy.orm import selectinload
+
+import src.util.datetime_util as datetime_util
+from src.adapters import db
+from src.adapters.db import flask_db
+from src.api.route_utils import raise_flask_error
+from src.auth.jwt_user_http_token_auth import JwtUserHttpTokenAuth
+from src.db.models.user_models import UserApiKey
+from src.logging.flask_logger import add_extra_data_to_current_request_logs
+
+logger = logging.getLogger(__name__)
+
+# Initialize the authorization context for API Gateway key authentication
+# This uses the X-API-Key header which is the standard header that AWS API Gateway
+# forwards when api_key_required is set to true
+api_gateway_key_auth = JwtUserHttpTokenAuth(
+    "ApiKey", header="X-API-Key", security_scheme_name="ApiGatewayKeyAuth"
+)
+
+
+class ApiKeyValidationError(Exception):
+    """Exception raised when API key validation fails"""
+
+    def __init__(self, message: str):
+        self.message = message
+        super().__init__(message)
+
+
+@api_gateway_key_auth.verify_token
+@flask_db.with_db_session()
+def verify_api_key(db_session: db.Session, token: str) -> UserApiKey:
+    logger.info("Authenticating API Gateway key")
+
+    try:
+        api_key = validate_api_key_in_db(token, db_session)
+
+        # Update last_used timestamp
+        api_key.last_used = datetime_util.utcnow()
+        db_session.add(api_key)
+        # Note: We don't commit here as the transaction will be managed
+        # by the calling context (e.g., @flask_db.with_db_session)
```

Comment:
We should commit here. There isn't actually any logic we've built into the `with_db_session` that does anything to close the transaction and while I think SQLAlchemy does do that when the session isn't being used anymore, I've seen inconsistencies in how that works. Also, if there were any issues with doing the commit, we'd rather the stack trace be in our code and not 12 layers deep in SQLAlchemy code.

Just add a:

```py
with db_session.begin():
     try:
         ...
```
around this section of code.